File: lectures/week1/files2.py
import os
import sys

# Open a file, write something to it and then close it
#
# Uses the `os` module to directly deal with integer file descriptors and the
# raw open(2), write(2) and close(2) system calls.

fd = os.open("output.txt", os.O_WRONLY | os.O_CREAT)

if fd < 0:
    # Negative return values indicate an error
    print("Cannot open output.txt for writing")
    sys.exit(-1)

count = os.write(fd, b"Hello world!")

# TODO: Unless `count == len("Hello World!")`, not all of the bytes we were
# expecting to be written necessarily have been.

os.close(fd)

File: lectures/week1/files2.c
#include <stdio.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

int main(void) {
    int fd;
    int count;

    fd = open("output.txt", O_WRONLY | O_CREAT);

    if (fd < 0) {
        printf("Cannot open output.txt for writing\n");
        perror("open");
        return -1;
    }

    count = write(fd, "Hello world!", sizeof ("Hello world!") - 1);

    /* TODO: Unless `count == strlen("Hello World!")`, not all of the bytes we were
     * expecting to be written necessarily have been. */
    if (count < 0) {
        perror("write");
        return -1;
    }

    if (close(fd) < 0) {
        perror("close");
        return -1;
    }

    return 0;
}

